### ISSUE TITLE
Provide a way to add a friendly obstruction when open measurement is used

### ISSUE THREAD

**Issue by vmagaziy** [appnexus/mobile-sdk-ios#49]
When articles containing ads are rendered using `UITextView`, they are to be covered by specific supplementary views to provide continuous selection:

![Simulator Screen Shot - iPhone X - 2020-03-11 at 13 57 11](https://user-images.githubusercontent.com/1112653/76419227-49112900-63a0-11ea-9ffd-19c6632f585c.png)

<img width="500" alt="Screenshot 2020-03-11 at 14 06 09" src="https://user-images.githubusercontent.com/1112653/76421224-a5c21300-63a3-11ea-82af-e27ece93bc83.png">

But putting ads behind these aux views breaks a viewability impression tracking made by [Open Measurement SDK](https://iabtechlab.com/standards/open-measurement-sdk/).

Fortunately, there is a way to [register obstructions](https://interactiveadvertisingbureau.github.io/Open-Measurement-SDKiOS/#register-obstructions) in the SDK, but it has to be exposed by the `AppNexusSDK`.

One of possible solutions is to introduce `- [ANBannerAdView addOpenMeasurementFriendlyObstruction:]` like shown in the following patch:

```
diff --git a/sdk/sourcefiles/ANBannerAdView.h b/sdk/sourcefiles/ANBannerAdView.h
index 390e6717..31bbe062 100644
--- a/sdk/sourcefiles/ANBannerAdView.h
+++ b/sdk/sourcefiles/ANBannerAdView.h
@@ -238,6 +238,12 @@ typedef NS_ENUM(NSUInteger, ANBannerViewAdAlignment) {
  */
 @property (nonatomic, readwrite) BOOL enableNativeRendering;
 
+/**
+ *  Excludes a given view from ad viewability calculations
+ *  when open measurement is used.
+ */
+- (void)addOpenMeasurementFriendlyObstruction:(nonnull UIView *)view;
+
 
 /**
  *  nativeAdRendererId :  Native Assembly renderer_id that is associated with the placement.
diff --git a/sdk/sourcefiles/Viewability/ANOMIDImplementation.h b/sdk/sourcefiles/Viewability/ANOMIDImplementation.h
index 222bd75d..8b72d6b2 100644
--- a/sdk/sourcefiles/Viewability/ANOMIDImplementation.h
+++ b/sdk/sourcefiles/Viewability/ANOMIDImplementation.h
@@ -33,5 +33,7 @@
 - (OMIDAppnexusAdSession*) createOMIDAdSessionforNative:(UIView *)view withScript:(NSMutableArray *)scripts;
 - (void) stopOMIDAdSession:(OMIDAppnexusAdSession*) omidAdSession;
 - (void)fireOMIDImpressionOccuredEvent:(OMIDAppnexusAdSession*) omidAdSession;
+- (void)addFriendlyObstruction:(UIView *) view
+    toOMIDAdSession:(OMIDAppnexusAdSession*) omidAdSession;
 
 @end
diff --git a/sdk/sourcefiles/internal/ANBannerAdView.m b/sdk/sourcefiles/internal/ANBannerAdView.m
index d082c88b..f7f2b631 100644
--- a/sdk/sourcefiles/internal/ANBannerAdView.m
+++ b/sdk/sourcefiles/internal/ANBannerAdView.m
@@ -281,8 +281,15 @@ - (NSTimeInterval)autoRefreshInterval {
     return __autoRefreshInterval;
 }
 
-
-
+- (void)addOpenMeasurementFriendlyObstruction:(nonnull UIView *)view {
+    if ([self.contentView isKindOfClass:[ANMRAIDContainerView class]])
+    {
+        ANMRAIDContainerView *standardAdView = (ANMRAIDContainerView *)self.contentView;
+        if (standardAdView.webViewController.omidAdSession != nil) {
+            [[ANOMIDImplementation sharedInstance] addFriendlyObstruction:view toOMIDAdSession:standardAdView.webViewController.omidAdSession];
+        }
+    }
+}
 
 #pragma mark - Transitions
 

```

**Comment by vmagaziy** [comment]
Native ads are to be supported as well, so `ANNativeAdResponse` should be extended to add obstructions to underlying session too.

**Comment by vmagaziy** [comment]
I see a corresponding API is in 7.3, but unfortunately it does not work for native ads.

When native ad response is received (`ANBannerAdView.m:496`) it has to add friendly obstruction views from the `obstructionViews` property to `ANNativeAdResponse.omidAdSession` as well.

Ping @asharmaa 🙇

**Comment by Kowshickkarthick** [comment]
HI @vmagaziy,

1) For `ANNativeAdResponse` received via `ANBannerAdView`/`ANNativeAdRequest` you need register Friendly obstruction view using the below API which has been newly added.

`- (BOOL)registerViewForTracking:(nonnull UIView *)view
         withRootViewController:(nonnull UIViewController *)rvc
                 clickableViews:(nullable NSArray<UIView *> *)views
                 openMeasurementFriendlyObstructions:(nonnull NSArray<UIView *> *)obstructionViews
                          error:(NSError *__nullable*__nullable)error;
`

`ANNativeAdResponse.h:200`
https://github.com/appnexus/mobile-sdk-ios/blob/master/sdk/sourcefiles/native/ANNativeAdResponse.h


2) We apply the friendly obstructions automatically if the `ANNativeAdResponse`  is rendered by the SDK using Native Assembly. (ANBannerAdView.m:448)

**Comment by vmagaziy** [comment]
@Kowshickkarthick, unfortunately usage of the proposed API is not applicable - registration is performed once but obstruction views can be added later on.

**Comment by Kowshickkarthick** [comment]
Hi @vmagaziy ,

For Native Ad's `registerViewForTracking` is the final method before the sdk starts tracking for Ad's viewability and impression.

If any of the parameters used during the first invocation of `registerViewForTracking` method has changed and does not apply anymore, your app needs to re-register the ad by calling `registerViewForTracking` again with the updated params.

This is the correct way to update.

**Comment by ppuviarasu** [comment]
Implemented & released with v7.3
